OPEN_PROBLEMS/HETEROGENEOUS_RUNTIMES.md

```diff
@@ -0,0 +1,190 @@
+
+#### What defines a complete solution?
+-   Assume an environment where several mobile devices run the same application and the application content between devices needs to be synchronised at all times. When one of the devices produces new content, the content should propagate to the rest of the devices realising a "push" model.
+-   Setup an environment with a sample application that produces content every so often.
+-   The devices can either be permanently connected in a mesh, or connect on demand and transfer content to each other.
+-   As nodes propagate content, they have to optimise the amount of data they exchange. Instead of just flooding full messages, nodes should first check whether their neighbours already have the latest content items. This can be achieved through CRDTs, or bloom-filters. At least one should be demonstrated.
+
+#### References: 
+-   Improve offline support: <https://github.com/libp2p/notes/issues/16>
+-   Offline message queue: <https://github.com/libp2p/notes/issues/2>
+-   IPFS CRDT datastore: <https://github.com/ipfs/go-ds-crdt>
+-   Optimize Circuit Relays to make it packet-oriented in libp2p: <https://github.com/libp2p/go-libp2p/issues/1019>
+-   NAT Traversal effors around libp2p: <https://github.com/libp2p/go-libp2p/issues?q=is%3Aissue+is%3Aopen+NAT+Traversal>
+
+### 🛠️ (Bonus) Area V: Extensibility / P2P VM a.k.a InterPlanetary Runtime
+This area of research is still a work in progress, and it may becoming a full-fledge open problem of its own, but for now we still consider it as an area inside the heterogeneous runtime problem.
+Maintaining and growing libp2p protocols for the gamut of runtimes and connections we want to support can be cumbersome. This is why we should consider embedding a common execution framework to every libp2p implementation so libp2p modules are compiled targeting this common framework and can be thus used in any of the runtimes. With this, protocols and modules are implemented once and used anywhere.
+
+Not only protocol implementations can benefit from embedding an *"InterPlanetary Runtime* in libp2p nodes, but also any other standard computation in the ecosystem such as IPLD codecs. IPLD codecs can be implemented targeting the *IP Runtime* so they don't have to also be implemented for every libp2p implementation and runtime. This opens the door to future lines of research such as *self-describing blocks*, where the block descriptions embeds the type of data and a pointer to the codec (which can be another block in stored in the network); *content-addressable code executions*, where code snippets are stored in the network, and data is specified in the code as pointers to blocks in th network; *computation delegation*, as once every node in the network shares the same common runtime and has a way to retrieve code and data directly from the network with a standard interface, delegating the execution of code to other's is as easy as sharing a link to the code and the data.
```

Comment:
This could likely go into the IPLD Open Problem (which hasn't been written yet). It can help on the interoperability, but seems a non priority for this Open Problem